pkg/proxy/service.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	proxyutil "k8s.io/kubernetes/pkg/proxy/util"
+)
+
+// ServiceChangeMap maps a service to its serviceChange.
+type ServiceChangeMap struct {
+	// lock protects items.
+	lock sync.Mutex
+	// items is a service to its serviceChange map.
+	items map[types.NamespacedName]*serviceChange
+}
+
+// serviceChange contains all changes to services that happened since proxy rules was synced.  For a single object,
+// changes are accumulated, i.e. previous is state from before all of them,
+// current is state after applying all of those.
```

Comment:
Nit: "i.e. previous is the state before changes, current is the state after changes"?

Reply:
Done.